lib/spack/spack/externals.py

```diff
@@ -0,0 +1,425 @@
+                    )
+                elif len(candidates) > 1:
+                    candidates_str = (
+                        f" [candidates are {', '.join([str(x.spec) for x in candidates])}]"
+                    )
+                    raise ExternalDependencyError(
+                        f"the spec '{spec_str}' depends on '{query_spec}', but there are multiple "
+                        f"external specs that could satisfy the request{candidates_str}{line_info}"
+                    )
+
+                dependency_dict["id"] = candidates[0].config["id"]
+
+    def _parse_all_nodes(self) -> None:
+        """Parses all the nodes from the external dicts but doesn't add any edge."""
+        for external_dict in self.external_dicts:
+            line_info = _line_info(external_dict)
+            try:
+                node = node_from_dict(external_dict)
+            except spack.spec.UnsatisfiableArchitectureSpecError:
+                spec_str, target_str = external_dict["spec"], external_dict["required_target"]
+                tty.debug(
+                    f"[{__name__}]{line_info} Skipping external spec '{spec_str}' because it "
+                    f"cannot be constrained with the required target '{target_str}'."
+                )
+                continue
+            except ExternalSpecError as e:
+                warnings.warn(f"{e}{line_info}")
+                continue
+
+            package_exists = spack.repo.PATH.exists(node.name)
+
+            # If we allow non-existing packages, just continue
+            if not package_exists and self.allow_nonexisting:
+                continue
+
+            if not package_exists and not self.allow_nonexisting:
+                raise ExternalSpecError(f"Package '{node.name}' does not exist{line_info}")
+
+            eid = external_dict.setdefault("id", str(uuid.uuid4()))
+            if eid in self.specs_by_external_id:
+                other_node = self.specs_by_external_id[eid]
+                other_line_info = _line_info(other_node.config)
+                raise DuplicateExternalError(
+                    f"Specs {node} and {other_node.spec} cannot have the same external id {eid}"
+                    f"{line_info}{other_line_info}"
+                )
+
+            self.complete_node(node)
+
+            # Add a Python dependency to Python extensions that don't specify it
+            pkg_class = spack.repo.PATH.get_pkg_class(node.name)
+            if (
+                "dependencies" not in external_dict
+                and not node.dependencies()
+                and any([c.__name__ == "PythonExtension" for c in pkg_class.__mro__])
+            ):
+                warnings.warn(
+                    f"Spack is trying attach a Python dependency to '{node}'. This feature is "
+                    f"deprecated, and will be removed in v1.2. Please make the dependency "
+                    f"explicit in your configuration."
+                )
```

Comment:
I think this should say something more specific and tell them what to do, with file and line info, but it can be in some later PR.  Something like:

```suggestion
                warnings.warn(
                    f"{file}:{line}: Python extension {node} does not define a python dependency is "
                    f"guessing that it depends on {python_node} at {line_info}. Spack will stop guessing "
                    f"this dependency in version 1.2 and you will need to specify it explicitly. You can add "
                    f"the following to your `packages.yaml` to do this:"
                    <python extension YAML augmented with dep with suitably unique python spec>
                )
```

I actually think 1.2 is too soon to remove this -- I think we need a better solution (e.g. with detection or some kind of warning when packages that really need a particular type of dependency don't have it).  I think we should come up with something in the package API, e.g., make extensions require their extendee as a dep in `packages.yaml` AND do a good job of detecting it, before we remove this.